EKF/covariance.cpp

```diff
@@ -146,37 +146,42 @@ void Ekf::predictCovariance()
 	const float d_vel_bias_sig = dt * dt * math::constrain(_params.accel_bias_p_noise, 0.0f, 1.0f);
 
 	// inhibit learning of imu accel bias if the manoeuvre levels are too high to protect against the effect of sensor nonlinearities or bad accel data is detected
+	// xy accel bias learning is also disabled on ground as those states are poorly observable when perpendicular to the gravity vector
 	const float alpha = math::constrain((dt / _params.acc_bias_learn_tc), 0.0f, 1.0f);
 	const float beta = 1.0f - alpha;
 	_ang_rate_magnitude_filt = fmaxf(dt_inv * _imu_sample_delayed.delta_ang.norm(), beta * _ang_rate_magnitude_filt);
 	_accel_magnitude_filt = fmaxf(dt_inv * _imu_sample_delayed.delta_vel.norm(), beta * _accel_magnitude_filt);
 	_accel_vec_filt = alpha * dt_inv * _imu_sample_delayed.delta_vel + beta * _accel_vec_filt;
 
-	if (_ang_rate_magnitude_filt > _params.acc_bias_learn_gyr_lim
-	    || _accel_magnitude_filt > _params.acc_bias_learn_acc_lim
-	    || _bad_vert_accel_detected) {
+	const bool is_maneuvre_level_high = _ang_rate_magnitude_filt > _params.acc_bias_learn_gyr_lim
+					    || _accel_magnitude_filt > _params.acc_bias_learn_acc_lim;
 
-		// store the bias state variances to be reinstated later
-		if (!_accel_bias_inhibit) {
-			_prev_dvel_bias_var = P.slice<3,3>(13,13).diag();
-		}
+	const bool do_inhibit_all_axes = (_params.fusion_mode & MASK_INHIBIT_ACC_BIAS)
+				   	 || is_maneuvre_level_high
+					 || _bad_vert_accel_detected;
 
-		_accel_bias_inhibit = true;
+	for (unsigned stateIndex = 13; stateIndex <= 15; stateIndex++) {
+		const int index = stateIndex - 13;
 
-	} else {
-		if (_accel_bias_inhibit) {
-			// reinstate the bias state variances
-			P(13,13) = _prev_dvel_bias_var(0);
-			P(14,14) = _prev_dvel_bias_var(1);
-			P(15,15) = _prev_dvel_bias_var(2);
+		// TODO: When on ground, only consider an accel bias observable if aligned with the gravity vector
+		// const bool is_bias_observable = (fabsf(_R_to_earth(2, index)) > 0.8f) || _control_status.flags.in_air;
+		const bool is_bias_observable = true;
+		const bool do_inhibit_axis = do_inhibit_all_axes || !is_bias_observable;
 
-		} else {
+		if (do_inhibit_axis) {
 			// store the bias state variances to be reinstated later
-			_prev_dvel_bias_var = P.slice<3,3>(13,13).diag();
```

Comment:
This case is missing now and I can't figure out if it matters.

Reply:
@kamilritz I think it's actually nicer the way it's done now, because you only save the previous variances once you start inhibiting. Previously from what I understand it was updates each cycle when biases were not inhibited.

Reply:
Yes, that's why I removed it during the refactoring. It's only used after passing through the other path that stores it as well.

Reply:
Agree